_posts/2021-03-10-dask_memory_usage.md

```diff
@@ -0,0 +1,261 @@
+If I had 12 inputs and each processing step took about the same time, they might run as follows:
+
+* CPU0: `f(X0)`, `f(X4)`, `f(X8)`
+* CPU1: `f(X1)`, `f(X5)`, `f(X9)`
+* CPU2: `f(X2)`, `f(X6)`, `f(X10)`
+* CPU3: `f(X3)`, `f(X7)`, `f(X11)`
+
+CPU is not the only bottleneck however, there is also memory to take into account, and this is where the problems started.
+
+Let's say each call to `f(Xi)` takes 12GB of RAM.
+That means to fully utilize 4 CPUs, I would need 48GB of RAM—but what if my computer only has 16GB of RAM?
+
+**Even though my computer has 4 CPUs, I can only utilize one CPU on a computer with 16GB RAM, because I don't have enough RAM to run more than one task in parallel.**
+In practice, these tasks ran in the cloud, where I could ensure the necessary RAM/core ratio was preserved by choosing the right pre-configured VM instances.
+And on some clouds you can freely set the amount of RAM and number of CPU cores for each virtual machine you spin up.
+
+But in order to provision hardware appropriately, I needed to know how much peak memory each task was using.
+
+## Introducing `dask-memusage`
+
+`dask-memusage` is a tool for measuring peak memory usage for each task in the Dask execution graph.
+
+* Per *task* because Dask executes code as a graph of tasks, and the graph determines how much parallelism can be used.
+* *Peak* memory is important, because that is the bottleneck.
+  It doesn't matter if average memory usage per task is 4GB, if two tasks need 12GB each at the same time your computer needs 24GB of RAM to be available.
+
+The interaction of the task graph and parallelism best illustrated with an example.
+Let's consider some trivial code that squares some numbers and then sums them:
+
+```python
+from dask.bag import from_sequence
+
+def square(x):
+    return x * x
+
+result = from_sequence(range(5)).map(square).sum()
+```
+
+We can visualize the task graph, the series of steps that Dask will try to execute:
+
+```python
+result.visualize(filename="example1.png", rankdir="TD")
+```
+
+Here's what it looks like:
+
+<img src="/images/dask_memusage/example1.png"  style="width: 75%; margin: 2em;">
+
+As we'd expect, squaring can happen in parallel, but summing the resulting values eventually devolves to a single action without parallelism because eventually we're combining down to a single result.
+
+For memory use purposes, in this particular example the peak memory usage would worst case be the peak memory usage of `square()` multiplied by the level of parallelism.
+So in the highly unlikely event that squaring a number takes 1GB of RAM, we'd need 10GB of RAM if we ran 10 `square()` calls in parallel.
+
+# Using `dask-memusage`
+
+Let's try a slightly more realistic example: we're going to count the occurrence of words in some text files, and then report the top-10 most common words in each file.
```

Comment:
Maybe "smaller" is a better word than "realistic" here? If anything, I think your first example is more realistic (because it is a real world scenario)

Reply:
Upon re-reading: are you using "realistic" to compare the word counting example with the squaring example? It's really not clear to me what you meant here.

Reply:
Realistic compared to squaring numbers, which is fairly trivial. But I think I'll remove the whole squaring thing and just assume readers know about task graph.